DSPy flavor docs

michael-berk opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/michael-berk/mlflow/pull/13289?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13289/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13289
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?
Done
* Landing page structure
* Landing page text
* Quickstart notebook
* MLflow integration architecture image
* (Model logging image
* FAQ section
* Add DSPy card to llm landing page

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [x] Examples
  - [x] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

DSPy flavor documentation.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

michael-berk committed:
DSPy landing page and structure

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Quickstart and images

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
DSPy Architecture

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Try to compile 1

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Update FAQ and toctree

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Update rendering issues and add flavor card

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
fix extension

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

chenmoneygithub commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,255 @@
+exploration of complex pipelines, allowing for fine-tuning performance based on specific tasks or 
+nuanced metrics. 
+
+.. figure:: ../../_static/images/llms/dspy/dspy-integration-architecture.png
```
We don't have tracking yet. IMO it's okay to use this image for a preview, but let's put up a disclaimer that DSPy tracking is not ready yet, will be released in a future version.

michael-berk replied:
@BenWilson2 see [this](https://github.com/mlflow/mlflow/pull/13289#discussion_r1796057394) comment

chenmoneygithub commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,255 @@
+
+The native integration of the DSPy library with MLflow helps users manage the development lifecycle with DSPy. The following are some of the key benefits of using DSPy with MLflow:
+
+* `MLflow Tracking <../../tracking.html>`_ allows you to track your DSPy program's training and execution. With the MLflow APIs, you can log a variety of artifacts and organize training runs, thereby increasing visibility into your model performance.  
```
we don't have this now, let's at least add a note here. 

@BenWilson2  Ben do you have some suggestions here?

BenWilson2 replied:
I'm a bit confused. 
We have `log_model()`, yes? 
I think the paragraph in question here is accurate with what is available in MLflow for the 2.18.0 release.

BenWilson2 replied:
If we're talking about autolog functionality, then yes, that isn't available, but we don't need to mention that. When that is ready, a new page will be needed for the docs and it can be linked in a bullet point section below this.

michael-berk replied:
Aligned. @chenmoneygithub does this look reasonable to you?

chenmoneygithub commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,255 @@
+
+    notebooks/dspy_quickstart.ipynb
+
+Concepts
```
We probably don't need this section, and in fact it's pretty hard to summarize DSPy components accurately in a few sentences. I would prefer linking them to https://dspy-docs.vercel.app/docs/category/dspy-building-blocks

michael-berk replied:
Good call I added links to the descriptions

chenmoneygithub commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,255 @@
+When using :py:func:`mlflow.dspy.log_model` or :py:func:`mlflow.dspy.save_model` in MLflow, the 
+DSPy program is serialized and saved to the tracking server as a ``.pkl`` file. This enables easy 
+deployment. Under the hood, MLflow uses ``cloudpickle`` to serialize the DSPy object, but not all 
+DSPy artifacts are serializable. 
```
we can delete "but not all DSPy artifacts are serializable", as it's unclear to users what's not serializable

michael-berk replied:
They're listed below, no?

BenWilson2 replied:
Looks like a typo? "but DSPy artifacts are serializable" <- let's fix that

chenmoneygithub commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,255 @@
+
+    # Define a model signature
+    input_schema = Schema([ColSpec("string")])
+    output_schema = Schema([ColSpec("string")])
```
this is my bad, but the `output_schema` is not correct here. In fact manually setting a dspy output schema is pretty hard, let's just delete this and let it automatically infer the schema from `input_example`

chenmoneygithub commented on docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb:
```diff
@@ -0,0 +1,1265 @@
+    "\n",
+    "The process follows these three steps, common to most DSPy [optimizers](https://dspy-docs.vercel.app/docs/building-blocks/optimizers):\n",
+    "\n",
+    "1. **Candidate Generation**: DSPy finds all `Predict` modules in the program and generates variations of instructions and demonstrations (e.g., examples for prompts). This step creates a set of possible candidates for the next stage.\n",
```
"instructions and demonstrations" => "instructions or demonstrations", as many optimizers won't touch "instructions".

chenmoneygithub commented on docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb:
```diff
@@ -0,0 +1,1265 @@
+    "The process follows these three steps, common to most DSPy [optimizers](https://dspy-docs.vercel.app/docs/building-blocks/optimizers):\n",
+    "\n",
+    "1. **Candidate Generation**: DSPy finds all `Predict` modules in the program and generates variations of instructions and demonstrations (e.g., examples for prompts). This step creates a set of possible candidates for the next stage.\n",
+    "2. **Parameter Optimization**: DSPy then uses methods like random search, TPE, or Optuna to select the best configurations. Fine-tuning models can also be done at this stage.\n",
```
best configurations => best candidate for consistency with step 1. 

Finetuning is an extra step that doesn't fit here, we can just remove this sentence to avoid confusion.

michael-berk replied:
Good call on wording!

Regarding fine tuning, isn't this a [fine tuning optimizer](https://dspy-docs.vercel.app/api/optimizers/BootstrapFinetune)? I know we chatted about it not working due to bugs, but the package does in theory support this.

chenmoneygithub commented on docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb:
```diff
@@ -0,0 +1,1265 @@
+    "\n",
+    "1. **Candidate Generation**: DSPy finds all `Predict` modules in the program and generates variations of instructions and demonstrations (e.g., examples for prompts). This step creates a set of possible candidates for the next stage.\n",
+    "2. **Parameter Optimization**: DSPy then uses methods like random search, TPE, or Optuna to select the best configurations. Fine-tuning models can also be done at this stage.\n",
+    "3. **Higher-Order Optimization**: Lastly, DSPy can modify the program's structure, such as creating ensembles of different variations of the program to boost performance.\n",
```
hmm which optimizer does this?

michael-berk replied:
Great callout. According to [this paper](https://arxiv.org/pdf/2406.11695) temperature on the `proposer LM`, not the optimized LM. I misinterpreted and will fix.

```
0-Shot MIPRO++ tunes instructions by metaoptimizing how or whether our Grounded instruction proposal strategy uses the dataset summary
(boolean), uses the program summary (boolean),
adjusts the proposer LM temperature (float), provides the proposer LM a plain-text tip on prompt
engineering (categorical; see Appendix C), and
selects a specific set of bootstrapped demos to show
to the proposer LM (categorical). A Bayesian
model with the same mini-batching strategy employed in MIPRO is then used to optimize over
these hyperparameters. A program with the newly
proposed instruction is evaluated on each trial, using the mini-batching approach described earlier.
The best fully-evaluated program is returned.
```

chenmoneygithub commented on docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb:
```diff
@@ -0,0 +1,1265 @@
+    }
+   ],
+   "source": [
+    "%pip install -U openai dspy-ai>=0.2.5 mlflow>=2.17.0"
```
let's use `dspy`, since 2.5.1 `dspy` becomes the golden package name instead of `dspy-ai`

chenmoneygithub commented on docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb:
```diff
@@ -0,0 +1,1265 @@
+    }
+   ],
+   "source": [
+    "lm.inspect_history(n=1)"
```
this is useful for debugging, but pretty long and hard to interpret for new users, we can delete this chunk.

chenmoneygithub commented on docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb:
```diff
@@ -0,0 +1,1265 @@
+    "from mlflow.types.schema import ColSpec, Schema\n",
+    "\n",
+    "input_schema = Schema([ColSpec(\"string\")])\n",
+    "output_schema = Schema([ColSpec(\"string\")])\n",
```
same here - let's delete the schema definition, and let mlflow infer from the input_example

chenmoneygithub reviewed (changes requested):
Awesome work!

michael-berk committed:
PR comments

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+The native integration of the DSPy library with MLflow helps users manage the development lifecycle with DSPy. The following are some of the key benefits of using DSPy with MLflow:
+
+* `MLflow Tracking <../../tracking.html>`_ allows you to track your DSPy program's training and execution. With the MLflow APIs, you can log a variety of artifacts and organize training runs, thereby increasing visibility into your model performance.  
+* `MLflow Model <../../models.html>`_ packages your compiled DSPy program with all its dependency versions, input and output interfaces, and other essential metadata. This allows you to deploy your compiled DSPy program with ease, knowing that the environment is consistent across different stages of the ML lifecycle.
```
```suggestion
* `MLflow Model <../../models.html>`_ packages your compiled DSPy program along with  its dependency versions, input and output interfaces and other essential metadata. This allows you to deploy your compiled DSPy program with ease, knowing that the environment is consistent across different stages of the ML lifecycle.
```

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+
+.. code-block:: shell
+
+    pip install mlflow dspy -U
+.. code-block:: python
```
```suggestion
    pip install mlflow dspy -U

.. code-block:: python
```

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+
+.. tip::
+
+    Typically you'd want to leverage a compiled DSPy module. MLflow will natively supports logging
```
Can we link to the DSPy docs about module compilation?

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+Logging the Program to MLflow
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can log the ``dspy.Module`` object to the MLflow experiment using the 
```
```suggestion
You can log the ``dspy.Module`` object to an MLflow run using the 
```

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+You can log the ``dspy.Module`` object to the MLflow experiment using the 
+:py:func:`mlflow.dspy.log_model` function.
+
+We will also specify a model signature. An MLflow model signature defines the expected schema for 
```
Can we link to the signatures documentation here, specifically the section on auto-signature generation using input examples?

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+deployment. Under the hood, MLflow uses ``cloudpickle`` to serialize the DSPy object, but 
+DSPy artifacts are serializable. 
+
+Here are some relevant serialized artifacts that help you reproduce a DSPy program in another 
+environment.
+* A pickle file containing the DSPy program artifact
+* The DSPy settings, which are serialized in the above ``.pkl`` file
+* An MLflow model signature
+* A pip requirements file containing inferred or user-specified pip dependencies
+* `venv` and `conda` environment files.
+
+Here are some relevant objects that are **not** serialized and need to be handled by the user. 
+* API tokens. These should be managed separately and passed securely via environment variables.
+* The DSPy trace object, which is primarily used during training, not inference.
```
I don't think we need to mention the aspects that are serialized and how they help (it's fairly obvious :) ). A list of objects that are not included in the serialization operation for DSPy **is** useful, however!

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+When serializing using the MLflow DSPy flavor, tokens are dropped from the settings objects. It is
+the user responsibility to securely pass the required secrets to the deployment environment. 
```
```suggestion
the user's responsibility to securely pass the required secrets to the deployment environment. 
```

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+How is the DSPy ``settings`` object saved?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To ensure program reproducibility, the service context is convert to a Python dictionary and 
```
```suggestion
To ensure program reproducibility, the service context is converted to a Python dictionary and 
```

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To ensure program reproducibility, the service context is convert to a Python dictionary and 
+pickled with the model artifact. This entire process is handled automatically behind the scenes.
```
```suggestion
pickled with the model artifact. 
```

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+To ensure program reproducibility, the service context is convert to a Python dictionary and 
+pickled with the model artifact. This entire process is handled automatically behind the scenes.
+
+Remember that your API keys are dropped.
```
```suggestion
Remember that your API keys are dropped.
```
```suggestion
Sensitive API access keys that are set within the ``Settings`` object are not persisted when logging your model. When deploying your DSPy model, you must ensure that the deployment environment has these keys set so that your DSPy model can make remote calls to services that require access keys.
```

BenWilson2 commented on docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb:
```diff
@@ -0,0 +1,927 @@
+    "1. Define a [dspy.Signature](https://dspy-docs.vercel.app/docs/building-blocks/signatures) and [dspy.Module](https://dspy-docs.vercel.app/docs/building-blocks/modules) to perform text classification.\n",
+    "2. Leverage [dspy.teleprompt.BootstrapFewShotWithRandomSearch](https://dspy-docs.vercel.app/api/optimizers/BootstrapFewShotWithRandomSearch) to compile our module so it's better at classifying our text.\n",
+    "3. Log the compiled model with MLflow.\n",
+    "4. Load the compiled model with MLflow and perform inference."
```
```suggestion
    "4. Load the logged model and perform inference."
```

BenWilson2 commented on docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb:
```diff
@@ -0,0 +1,927 @@
+   "source": [
+    "import dspy\n",
+    "\n",
+    "# Define your model. We will use OpenAI for simplicity\n",
```
? Are we using OpenAI or Llama? 
Either way is fine; let's just be consistent in the text.

BenWilson2 commented on docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb:
```diff
@@ -0,0 +1,927 @@
+   },
+   "source": [
+    "## 1.1 - Hello World\n",
+    "Let's demonstrate predicting via the DSPy module and associated signature. As expected, the program incorrectly classifies our text because it has not been given the labels."
```
Is this cheating? 
Can we at least compare the results between GPT-4 (as you're using) with a basic prompt that includes the label definitions vs. the evaluated results of the compiled and optimized program? 
Are the results compelling in that case?

michael-berk replied:
Yes accuracy is still improved with labels. I can refactor.
```
Uncompiled accuracy: 0.625
Compiled accuracy: 0.875
```
For context, the purpose of not providing the labels is to show that DSPy can infer the labels and incorporate them.

michael-berk committed:
Update docs/source/llms/dspy/index.rst

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Update docs/source/llms/dspy/index.rst

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Update docs/source/llms/dspy/index.rst

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Update docs/source/llms/dspy/index.rst

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Update docs/source/llms/dspy/notebooks/dspy_quickstart.ipynb

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Update docs/source/llms/dspy/index.rst

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Update docs/source/llms/dspy/index.rst

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Update docs/source/llms/dspy/index.rst

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Ben PR comments

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk committed:
Fix underline

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

chenmoneygithub reviewed (approved):
LGTM!

BenWilson2 commented:
This is going to be VERY annoying, but....
- Can the image at the top of the DSPy page that shows the architecture be rearranged so that it is legible? Perhaps vertically-based design so that the fonts are readable? You can see how small everything is: https://output.circle-artifacts.com/output/job/2cf4c02b-f38e-4ecb-9164-8a59183bcb50/artifacts/0/docs/build/html/llms/dspy/index.html 
- The tip block immediately above https://output.circle-artifacts.com/output/job/2cf4c02b-f38e-4ecb-9164-8a59183bcb50/artifacts/0/docs/build/html/llms/dspy/index.html#logging-the-program-to-mlflow has a broken link. 
- Can the "service context" be explained in the FAQ section https://output.circle-artifacts.com/output/job/2cf4c02b-f38e-4ecb-9164-8a59183bcb50/artifacts/0/docs/build/html/llms/dspy/index.html#logging-the-program-to-mlflow ? There isn't a mention of how settings correlates to a 'service context'. 
- Can we remove all of the numbering in https://output.circle-artifacts.com/output/job/2cf4c02b-f38e-4ecb-9164-8a59183bcb50/artifacts/0/docs/build/html/llms/dspy/index.html#logging-the-program-to-mlflow ? The h-levels are sufficient to show groupings and the prefaced numbering doesn't look like how we do any other tutorial. 
- What does "load out" mean? https://output.circle-artifacts.com/output/job/2cf4c02b-f38e-4ecb-9164-8a59183bcb50/artifacts/0/docs/build/html/llms/dspy/index.html#logging-the-program-to-mlflow

michael-berk committed:
Ben Pr comments 2

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

michael-berk commented:
> * context

* Done
* Fixed - good catch
* Clarified where it's used
* This links to the main page. I assume you mean the quickstart and have updated the .ipynb
* Fixed in .ipynb

michael-berk commented:
> This is going to be VERY annoying, but....
> 
> * Can the image at the top of the DSPy page that shows the architecture be rearranged so that it is legible? Perhaps vertically-based design so that the fonts are readable? You can see how small everything is: https://output.circle-artifacts.com/output/job/2cf4c02b-f38e-4ecb-9164-8a59183bcb50/artifacts/0/docs/build/html/llms/dspy/index.html
> * The tip block immediately above https://output.circle-artifacts.com/output/job/2cf4c02b-f38e-4ecb-9164-8a59183bcb50/artifacts/0/docs/build/html/llms/dspy/index.html#logging-the-program-to-mlflow has a broken link.
> * Can the "service context" be explained in the FAQ section https://output.circle-artifacts.com/output/job/2cf4c02b-f38e-4ecb-9164-8a59183bcb50/artifacts/0/docs/build/html/llms/dspy/index.html#logging-the-program-to-mlflow ? There isn't a mention of how settings correlates to a 'service context'.
> * Can we remove all of the numbering in https://output.circle-artifacts.com/output/job/2cf4c02b-f38e-4ecb-9164-8a59183bcb50/artifacts/0/docs/build/html/llms/dspy/index.html#logging-the-program-to-mlflow ? The h-levels are sufficient to show groupings and the prefaced numbering doesn't look like how we do any other tutorial.
> * What does "load out" mean? https://output.circle-artifacts.com/output/job/2cf4c02b-f38e-4ecb-9164-8a59183bcb50/artifacts/0/docs/build/html/llms/dspy/index.html#logging-the-program-to-mlflow

* Done
* Fixed - good catch
* Clarified where it's used
* This links to the main page. I assume you mean the quickstart and have updated the .ipynb
* Fixed in .ipynb

michael-berk committed:
Update settings definition

Signed-off-by: Michael Berk <michaelberk99@gmail.com>

BenWilson2 commented on docs/source/llms/dspy/index.rst:
```diff
@@ -0,0 +1,247 @@
+        intelligence, or adaptability.''',
+        'answer': '''There isn\'t a definitive answer, as it depends on the context. However, many 
+        people humorously refer to the American black bear or the polar bear when discussing 
+        "the best" kind of bear.'
```
```suggestion
        "the best" kind of bear.'''
```

BenWilson2 committed:
Update docs/source/llms/dspy/index.rst

Signed-off-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>

BenWilson2 reviewed (approved):
LGTM!

BenWilson2 commented:
/rr
